### PR TITLE
Replace old RBAC handling with new RBAC handling

### PR DESCRIPTION
**Pull request by knanao**
**What this PR does / why we need it**:
I fixed codegen plugin to address new RBAC instead of old RBAC.
In this version, codegen enables to handle of a new MethodOption, which is `(model.rbac)` instead of `(model.role)`, 

*`(model.role)` will be not able to be used in the next version.*

Usage:
```
// Define a specific RPC.
rpc RegisterPiped(RegisterPipedRequest) returns (RegisterPipedResponse) {
      option (model.rbac).resource = PIPED;
      option (model.rbac).action = CREATE;
}

// If you want to ignore a specific RPC, you better use `(model.rbac).ignored`.
rpc GetCommand(GetCommandRequest) returns (GetCommandResponse) {
       option (model.rbac).ignored = true;
}

// Auto-generated like this code
case "/grpc.service.webservice.WebService/RegisterPiped":
       return rbac.HasPermission(model.ProjectRBACResource_PIPED, model.ProjectRBACPolicy_CREATE)
case "/grpc.service.webservice.WebService/GetCommand":
       return true
```

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/model/project.go [hunk 1]
```diff
@@ -518,3 +518,33 @@ func (p *Project) DeleteRBACRole(name string) error {
 	}
 	return fmt.Errorf("%s role does nott exist", name)
 }
+
+func (p *ProjectRBACRole) HasPermission(typ ProjectRBACResource_ResourceType, action ProjectRBACPolicy_Action) bool {
```
Comment (knanao) [inline pkg/model/project.go:522]: This method is called by [this](https://github.com/pipe-cd/pipecd/pull/3868/files#diff-1c6241b37c87f74eead2ec6ce129dd88bd10cd8ba1ebd5daa2246becd00e0c47R87).

File: tool/codegen/protoc-gen-auth/main.go [hunk 2]
```diff
@@ -21,15 +22,19 @@ type FileParams struct {
 }
 
 type Method struct {
-	Name string // The name of the RPC
-	Role string // ADMIN,EDITOR,VIEWER
+	Name     string // The name of the RPC
+	Resource string // APPLICAION,DEPLOYMENT,EVENT,PIPED,DEPLOYMENT_CHAIN,PROJECT,API_KEY,INSIGHT
```
Comment (nghialv) [inline tool/codegen/protoc-gen-auth/main.go:26]:
```suggestion
	Resource string // APPLICATION,DEPLOYMENT,EVENT,PIPED,DEPLOYMENT_CHAIN,PROJECT,API_KEY,INSIGHT
```

**Comment by nghialv** [comment]
From the PR's description, I think we can merge all the options into a single category. How about this?

``` proto
// Define a specific RPC.
rpc RegisterPiped(RegisterPipedRequest) returns (RegisterPipedResponse) {
      option (model.rbac).resources = PIPED;
      option (model.rbac).actions = CREATE;
}

rpc GetCommand(GetCommandRequest) returns (GetCommandResponse) {
       option (model.rbac).ignored = true;
}
```

**Comment by knanao** [comment]
If we want to do like that, we have to define RBAC model and redefine `ResourceType` and `Action`.
I mean, I have to manage both of them when we add a new resource or action.

**Comment by knanao** [comment]
@nghialv 
This is really nice idea!
I created a new model RBAC and fixed the logic of auth-plugin parser.
Finally, we can use it like below.

```
// Define a specific RPC.
rpc RegisterPiped(RegisterPipedRequest) returns (RegisterPipedResponse) {
      option (model.rbac).resource = PIPED;
      option (model.rbac).action = CREATE;
}

// If you want to ignore a specific RPC, you better use `(model.rbac).ignored`.
rpc GetCommand(GetCommandRequest) returns (GetCommandResponse) {
       option (model.rbac).ignored = true;
}

// Auto-generated like this code
case "/grpc.service.webservice.WebService/RegisterPiped":
       return rbac.HasPermission(model.ProjectRBACResource_PIPED, model.ProjectRBACPolicy_CREATE)
case "/grpc.service.webservice.WebService/GetCommand":
       return true
```

File: tool/codegen/protoc-gen-auth/file.go [hunk 3]
```diff
@@ -43,17 +61,66 @@ func isViewer(r model.Role) bool {
...
+func (a *authorizer) getRBAC(ctx context.Context, projectID string) (*rbac, error) {
+	r, err := a.rbacCache.Get(projectID)
+	if err != nil {
+		project, err := a.projectStore.Get(ctx, projectID)
+		if err != nil {
+			return nil, err
+		}
+		r = &rbac{project.RbacRoles}
+		a.rbacCache.Put(projectID, r)
+	}
```
Comment (knanao) [inline tool/codegen/protoc-gen-auth/file.go:73]: In order to not spike requests, I adapted look-aside caching pattern.

**Review by nghialv — APPROVED** [review]
Looks fine.

**Review by khanhtc1202 — APPROVED** [review]
Here you go 👍

**Pull request merged**